src/sympc/protocol/falcon/falcon.py

```diff
@@ -451,3 +454,100 @@ def select_shares(x: MPCTensor, y: MPCTensor, b: MPCTensor) -> MPCTensor:
         z = x + (d * (y - x))
 
         return z
+
+    @staticmethod
+    def _random_prime_group(
+        session: Session, shape: Union[torch.Size, tuple]
+    ) -> MPCTensor:
+        """Computes shares of random number in Zp*.Zp* is the multiplicative group mod p.
+
+        Args:
+            session (Session): session to generate random shares for.
+            shape (Union[torch.Size,tuple]): shape of the random share to generate.
+
+        Returns:
+            share (MPCTensor): Returns shares of random number in group Zp*.
+
+        Zp* = {1,2..,p-1},where p is a prime number.
+        We use Euler's Theorem for verifying that random share is not zero.
+        It states that:
+        For a general modulus n
+        a^phi(n) = 1(mod n), if a is co prime to n.
```

Comment:
Q: Why do we need this?
Can't we simply generate random shares in ring size `p` on each party and when reconstructing they give back a random number?

Reply:
Good question, we do this to prevent the random share from being zero,

Reply:
I will include as comment also.

Reply:
We try effectively to sample from the distribution [1,p-1]

Reply:
We cannot restrict parties to sample in a specific manner(which is a security leak), we allow them to sample from domain [0,p-1] and check if the shares sum to zero.

Reply:
We do it mainly to mask the c values, if the random shares is zero, it reveals the d output implicitly on reconstruction.